include/Target/Btor/BtorToBtorIRTranslation.h

```diff
@@ -0,0 +1,190 @@
+
+#include "mlir/IR/OwningOpRef.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+
+#include <vector>
+#include <map>
+
+#include "btor2parser/btor2parser.h"
+
+namespace mlir {
+class MLIRContext;
+class ModuleOp;
+
+namespace btor {
+
+/// Deserializes the given Btor module and creates a MLIR ModuleOp
+/// in the given `context`. Makes use of btor2parser.
+
+class Deserialize {
+
+ public:
+///===----------------------------------------------------------------------===//
+/// Constructors and Destructors
+///===----------------------------------------------------------------------===//
+
+  Deserialize(MLIRContext *context, const std::string &s) : context(context), 
+    builder(OpBuilder(context)), unknownLoc(UnknownLoc::get(context)) {
+        modelFile = fopen(s.c_str(), "r");
+    }
+
+  ~Deserialize() {
+      if (model) {
+        btor2parser_delete(model);
+      }
+      if (modelFile) {
+        fclose(modelFile);
+      }
+  }
+
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+
+  std::vector<Btor2Line *> inputs;
+  std::vector<Btor2Line *> states;
+  std::vector<Btor2Line *> bads;
+  std::vector<Btor2Line *> inits;
+  std::vector<Btor2Line *> nexts;
+  std::vector<Btor2Line *> constraints;
+ 
+  std::map<int64_t, Btor2Line *> reachedLines;
+  
+  bool parseModelIsSuccessful();
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+  
+  std::map<int64_t, Value> cache;
+
+  OwningOpRef<FuncOp> buildInitFunction();
+  OwningOpRef<FuncOp> buildNextFunction();
+  
+ private: 
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  Btor2Parser *model = nullptr;
+  FILE *modelFile = nullptr;
+
+  void parseModelLine(Btor2Line *l);
+
+  void filterInits(){
+    size_t i = 0;
+    for (size_t j = 0, sz = inits.size(); j < sz; ++j) {
+        if (inits.at(j)) {
+            inits[i++] = inits.at(j);
+        }
+    }
+    inits.resize(i);
+  }
+
+  void filterNexts() {
+    size_t i = 0;
+    for (size_t j = 0, sz = nexts.size(); j < sz; ++j) {
+      if (nexts.at(j)) {
+        nexts[i++] = nexts.at(j);
+      }
+    }
+    nexts.resize(i);
+  }
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+
+  MLIRContext *context;
+  OpBuilder builder;
+  Location unknownLoc;
+  
+  void toOp(Btor2Line *line);
+  bool isValidChild(Btor2Line * line);
+  void createNegateLine(int64_t curAt, Value child);
```

Comment:
`Value` passed by value, is that intentional?

Reply:
This can be passed by reference. The value we are getting should already be on the heap since we allocate it into the map